# Patch was generated with diff -ru1 a/ b/
diff -ru1 a/hal/phydm/phydm_interface.c b/hal/phydm/phydm_interface.c
--- a/hal/phydm/phydm_interface.c	2017-12-22 02:45:46.000000000 +0000
+++ b/hal/phydm/phydm_interface.c	2019-09-29 13:40:43.868293129 +0000
@@ -640,3 +640,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer,
+	struct legacy_timer		*p_timer,
	u32			ms_delay
@@ -645,5 +645,5 @@
 #if (DM_ODM_SUPPORT_TYPE & ODM_AP)
-	mod_timer(p_timer, jiffies + RTL_MILISECONDS_TO_JIFFIES(ms_delay));
+	mod_timer(&p_timer->t, jiffies + RTL_MILISECONDS_TO_JIFFIES(ms_delay));
 #elif (DM_ODM_SUPPORT_TYPE & ODM_CE) && defined(DM_ODM_CE_MAC80211)
-	mod_timer(p_timer, jiffies + msecs_to_jiffies(ms_delay));
+	mod_timer(&p_timer->t, jiffies + msecs_to_jiffies(ms_delay));
 #elif (DM_ODM_SUPPORT_TYPE & ODM_CE)
@@ -660,3 +660,3 @@
	struct PHY_DM_STRUCT			*p_dm,
-	struct timer_list			*p_timer,
+	struct legacy_timer			*p_timer,
	void	*call_back_func,
@@ -691,3 +691,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer
+	struct legacy_timer		*p_timer
 )
@@ -710,3 +710,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer
+	struct legacy_timer		*p_timer
 )
diff -ru1 a/hal/phydm/phydm_interface.h b/hal/phydm/phydm_interface.h
--- a/hal/phydm/phydm_interface.h	2017-12-22 02:45:46.000000000 +0000
+++ b/hal/phydm/phydm_interface.h	2019-09-29 13:40:33.772293543 +0000
@@ -340,3 +340,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer,
+	struct legacy_timer		*p_timer,
	u32			ms_delay
@@ -347,3 +347,3 @@
	struct PHY_DM_STRUCT			*p_dm,
-	struct timer_list			*p_timer,
+	struct legacy_timer			*p_timer,
	void	*call_back_func,
@@ -356,3 +356,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer
+	struct legacy_timer		*p_timer
 );
@@ -362,3 +362,3 @@
	struct PHY_DM_STRUCT		*p_dm,
-	struct timer_list		*p_timer
+	struct legacy_timer		*p_timer
 );
diff -ru1 a/hal/phydm/phydm_types.h b/hal/phydm/phydm_types.h
--- a/hal/phydm/phydm_types.h	2017-12-22 02:45:46.000000000 +0000
+++ b/hal/phydm/phydm_types.h	2019-09-29 13:20:32.820342791 +0000
@@ -133,3 +133,3 @@

-	#define	timer_list	_RT_TIMER
+	#define	legacy_timer	_RT_TIMER
	
diff -ru1 a/include/osdep_service_linux.h b/include/osdep_service_linux.h
--- a/include/osdep_service_linux.h	2017-12-22 02:45:46.000000000 +0000
+++ b/include/osdep_service_linux.h	2019-09-29 13:29:59.036319572 +0000
@@ -162,3 +162,3 @@
 #endif
-typedef struct timer_list _timer;
+//typedef struct timer_list _timer;
 typedef struct completion _completion;
@@ -294,2 +294,15 @@

+typedef struct legacy_timer {
+	struct timer_list t;
+	void (*function)(unsigned long);
+	unsigned long data;
+} _timer;
+
+
+static void legacy_timer_func(struct timer_list *t)
+{
+	struct legacy_timer *lt = (struct legacy_timer*)t;
+	lt->function(lt->data);
+}
+
 __inline static void _init_timer(_timer *ptimer, _nic_hdl nic_hdl, void *pfunc, void *cntx)
@@ -299,3 +312,4 @@
	ptimer->data = (unsigned long)cntx;
-	init_timer(ptimer);
+//	init_timer(ptimer);
+	timer_setup(&ptimer->t, legacy_timer_func, 0);
 }
@@ -304,3 +318,3 @@
 {
-	mod_timer(ptimer , (jiffies + (delay_time * HZ / 1000)));
+	mod_timer(&ptimer->t , (jiffies + (delay_time * HZ / 1000)));
 }
@@ -309,3 +323,3 @@
 {
-	*bcancelled = del_timer_sync(ptimer) == 1 ? 1 : 0;
+	*bcancelled = del_timer_sync(&ptimer->t) == 1 ? 1 : 0;
 }
diff -ru1 a/os_dep/linux/os_intfs.c b/os_dep/linux/os_intfs.c
--- a/os_dep/linux/os_intfs.c	2017-12-22 02:45:46.000000000 +0000
+++ b/os_dep/linux/os_intfs.c	2019-09-29 11:49:11.480567570 +0000
@@ -1266,9 +1266,9 @@

-static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0)
-	, void *accel_priv
-	#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0)
+static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb
+	#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
+	  , void *accel_priv
+	#else
+				, struct net_device *sb_dev
+	 #endif
	, select_queue_fallback_t fallback
-	#endif
-#endif
 )
diff -ru1 a/os_dep/osdep_service.c b/os_dep/osdep_service.c
--- a/os_dep/osdep_service.c	2017-12-22 02:45:46.000000000 +0000
+++ b/os_dep/osdep_service.c	2019-09-29 11:58:07.804545576 +0000
@@ -2002,3 +2002,3 @@
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 1, 0))
-		rlen = __vfs_read(fp, buf + sum, len - sum, &fp->f_pos);
+		rlen = kernel_read(fp, buf + sum, len - sum, &fp->f_pos);
 #else
